### PR TITLE
Ops and kernels for reduce_slice_ops

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This is a PR for https://github.com/tensorflow/tensorflow/issues/7662

This PR implements `reduce_slice_sum`, `reduce_slice_prod`, `reduce_slice_max`, `reduce_slice_min`.

There are also tests for these new ops.

No gradients are implemented now, but I will start working on that after this PR is merged.

### CODE REVIEW

**Comment by girving** [comment]
This is implementable with current ops by doing a scan followed by a gather.  Is that not sufficient for your purposes?

**Comment by drpngx** [comment]
I guess @zasdfgbnm has a point on the #7662. @aselle @girving what do you think?

**Comment by girving** [comment]
@drpngx Yep, @zasdfgbnm and @aselle convinced me the op is reasonable, though I switched the reviewer to Andy.

**Comment by caisq** [comment]
Gentle ping @aselle ?

**Comment by aselle** [comment]
I'm sorry I'll take a thorough look today.

**Comment by aselle** [comment]
Before merging this, can we think about if this will ever be useful to extend:
1. to higher order tensors than matrices. 
2. to axes that are not 0 and also if there are multiple axes?
If those things make sense and we can define them well. I'd like to make this function have a signature in the final form of that concept, so that we do not need to change the API to add that functionality.

It seems simple that we could define partial products of axis=1 right now where we sum over the columns rather than rows at least. This suggests we should have an axis parameter.

File: tensorflow/core/ops/math_ops.cc
Comment (aselle) [inline tensorflow/core/ops/math_ops.cc]: <pre>
"index" -> `indices`
</pre>

File: tensorflow/core/ops/math_ops.cc
Comment (aselle) [inline tensorflow/core/ops/math_ops.cc]: Please use an example with different values in each column to make the example more illustrative.

File: tensorflow/core/ops/math_ops.cc
Comment (aselle) [inline tensorflow/core/ops/math_ops.cc]: It might be worth using pretty print to make the matrices multi line so they are more readable see array_ops.cc:884 "MatrixBandPart"

File: tensorflow/core/ops/math_ops.cc
Comment (aselle) [inline tensorflow/core/ops/math_ops.cc]: Perhaps we should call this family of ops 
PartialReduceSum, PartialReduceProd, PartialReduceMin, PartialReduceMax so it is analogous to reduce_min, reduce_max, ...

File: tensorflow/python/ops/math_ops_test.py
Comment (aselle) [inline tensorflow/python/ops/math_ops_test.py]: thanks for testing this case.

File: tensorflow/python/ops/math_ops_test.py
Comment (aselle) [inline tensorflow/python/ops/math_ops_test.py]: test scalars for indices and values?

**Review by aselle — COMMENTED** [review]
Not quite done reviewing yet...

**Comment by aselle** [comment]
@tensorflow-jenkins test this please

**Comment by zasdfgbnm** [comment]
Response to @aselle 
1. Higher order tensor is actually already supported now.
2. This is a great point, I think I will change the code to support this. It won't require too much work to do so.

File: tensorflow/python/ops/math_ops_test.py
Comment (zasdfgbnm) [inline tensorflow/python/ops/math_ops_test.py]: Do you mean the case where "data" and "indices" are scalars? In my shape function, I explicitly forbid this case. See math_ops_test.cc: 493-495 in this PR

File: tensorflow/core/ops/math_ops.cc
Comment (zasdfgbnm) [inline tensorflow/core/ops/math_ops.cc]: Yes, I didn't think too much when naming this. Do you think naming them as RangedReduceSum, RangedReduceProd, RangedReduceMin, RangedReduceMax would be even better? The word "partial" is too general and the ops I'm adding can not do all types of "partial" reduction (for example adding all elements with odd index)

Comment (aselle) [inline tensorflow/core/ops/math_ops.cc]: It almost is a ReduceSliceSum becaue it's like you are taking a slice in one dimension and reducing along it? no?

Comment (zasdfgbnm) [inline tensorflow/core/ops/math_ops.cc]: Yes, "Slice" sounds better. Let's call it ReduceSliceSum for now before we can think of any better name.

**Comment by zasdfgbnm** [comment]
When implementing the reduction along any axis, I find it helpful to flat both inner and outer dims in the shape. So I create a new method `flat_inner_outer_dims` for class `Tensor` in a separate PR https://github.com/tensorflow/tensorflow/pull/9452

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by zasdfgbnm** [comment]
@drpngx this hasn't been done (actually still a lot to do). I just merge with master in order to use merged changes I made.

**Comment by zasdfgbnm** [comment]
Before changing this to add support for user specified axis, I would like to add a `GetCuda3DLaunchConfig` to `tensorflow/core/util/cuda_kernel_helper.h`, in a separate PR. The newly added function will be used here.

**Comment by drpngx** [comment]
Sounds good.

**Comment by rmlarsen** [comment]
@zasdfgbnm any update on your progress with GetCuda3DLaunchConfig?

**Comment by zasdfgbnm** [comment]
@rmlarsen There is some part of the existing code that makes me not feeling well.
See https://github.com/tensorflow/tensorflow/issues/9937
Once everything is clear, I think it won't take too long to get a `GetCuda3DLaunchConfig`

**Comment by rmlarsen** [comment]
@zasdfgbnm OK, thanks for the update! I'll mark this as stalled for now.

**Comment by zasdfgbnm** [comment]
@aselle I think all my changes are done for this PR, please continue the review process. This op now allows users to specify the axis. I also added one test case to test when `axis != 0` and one test case to test when `rank > 2`. All these tests pass on my machine.

This PR relies on https://github.com/tensorflow/tensorflow/pull/10032 and https://github.com/tensorflow/tensorflow/pull/10068 that has not been merged into master branch yet. I manually merge them into this branch to make this work. So you don't have to worry about changes in `shape_inference.cc` and `cuda_kernel_helper.h` because they are reviewed separately. You may not want to merge this before these two dependencies are merged.

**Comment by zasdfgbnm** [comment]
By the way, this PR include a fix in `tensor.h` related to const tensors

**Comment by shoyer** [comment]
Just to note, #8954 adds `tf.repeat`, which provides gradients for `partial_sum` (and vice-versa). See discussion at https://github.com/tensorflow/tensorflow/pull/8954#discussion_r118302537.

**Comment by zasdfgbnm** [comment]
@shoyer The name of `partial_sum` has been suggested to change to `reduce_slice_sum`, see https://github.com/tensorflow/tensorflow/pull/9063#discussion_r113259298
The `tf.repeat` and `tf.reduce_slice_sum` are gradients of each other regardless of the API difference. But the current API of `tf.repeat` and `tf.reduce_slice_sum` makes it a little bit tricky to use directly:

- In order to use `tf.reduce_slice_sum` as gradient of `tf.repeat`, there must be another op to compute the `indices` that `tf.reduce_slice_sum` takes.
- In order to use `tf.repeat` as gradient of `tf.reduce_slice_sum`, there must be another op to reorder and merge some gradients, and compute the `repeats` argument that `tf.repeat` takes.

**Comment by zasdfgbnm** [comment]
Currently, `indices` can only be a `n*2` matrix where the first column is the starting index and the second column is the ending index. It would be nice to extend the `indices` argument of `tf.reduce_slice_***` here to support a vector input. In the case of vector input, the second column of row `i` will be treated as the first column of row  `i+1`. For example:
```python
indices = [0,5,10,15]
``` 

is equivalent to

```python
indices = [ [0,5],
            [5,10],
            [10,15],
            [15,+infinity]]
```
This extension would make `tf.reduce_slice_***` more useful. And with this extension, the gradient of `tf.repeat` would become a very simple like
```python
tf.reduce_slice_sum(grad, tf.scan(lambda a, x: a + x, repeats)))
```

@zycdragonball I think in your pull request, you can submit a `tf.repeat` without implementation of gradient and wait for this PR to be merged. What do you think?

**Comment by zasdfgbnm** [comment]
The extend is done

**Comment by zycdragonball** [comment]
@zasdfgbnm I'll leave the gradient of `tf.repeat` as it is now, and change it to `tf.reduce_slice_sum` once it is merged.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
@zasdfgbnm this doesn't seem to build. Could you fix and push again?

**Comment by zasdfgbnm** [comment]
@drpngx It should be fixed now.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
@aselle WDYT? If OK, then we can trigger an API review, or recommend it go to contrib.

**Comment by aselle** [comment]
I'd like to put these into contrib for now. The reason being that we have repeat going into contrib, and this is needed for the gradients of repeat() and repeat() is needed for gradient here. Once they are both in contrib,  the gradients should be added, and once things have stabilized we should do api review and bring them in.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by zasdfgbnm** [comment]
Not done yet

**Comment by zasdfgbnm** [comment]
done now. should be working

_Sent from my OnePlus ONEPLUS A3000 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by zasdfgbnm** [comment]
seems to be unrelated fails?

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by rmlarsen** [comment]
@zasdfgbnm It looks like you need to add your op to the Cmake build. 
https://ci.tensorflow.org/job/tensorflow-pr-win-cmake-py/3768/consoleFull

**Comment by zasdfgbnm** [comment]
@rmlarsen I added the cmake build rules. Hopefully it will work. I'm unable to test it on my machine because I'm running into this issue: https://github.com/tensorflow/tensorflow/issues/12018

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by rmlarsen** [comment]
@zasdfgbnm Thanks!

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
@zasdfgbnm Thanks a lot for this nice contribution!

File: tensorflow/contrib/reduce_slice_ops/python/ops/reduce_slice_ops.py
Comment (andrewharp) [inline tensorflow/contrib/reduce_slice_ops/python/ops/reduce_slice_ops.py]: This is giving me a problem with an internal build:
```
    reduce_slice_sum = _reduce_slice_ops.reduce_slice_sum
AttributeError: 'module' object has no attribute 'reduce_slice_sum'
```
Is there an extra step necessary somewhere to expose the ops completely?

Comment (zasdfgbnm) [inline tensorflow/contrib/reduce_slice_ops/python/ops/reduce_slice_ops.py]: Hi @andrewharp, I didn't see any problem on this line on my computer. But I do see another problem on somewhere else. I added some more commits to my branch. See the file changes in:
https://github.com/tensorflow/tensorflow/compare/master...zasdfgbnm:partial_reduces

Do you think this will fix the problem you are talking about?

**Pull request merged**
